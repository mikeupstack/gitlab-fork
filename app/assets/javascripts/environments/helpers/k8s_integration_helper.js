import { differenceInSeconds } from '~/lib/utils/datetime_utility';

export function generateServicePortsString(ports) {
  if (!ports?.length) return '';

  return ports
    .map((port) => {
      const nodePort = port.nodePort ? `:${port.nodePort}` : '';
      return `${port.port}${nodePort}/${port.protocol}`;
    })
    .join(', ');
}

export function getServiceAge(creationTimestamp) {
  if (!creationTimestamp) return '';

  const timeDifference = differenceInSeconds(new Date(creationTimestamp), new Date());

  const seconds = Math.floor(timeDifference);
  const minutes = Math.floor(seconds / 60) % 60;
  const hours = Math.floor(seconds / 60 / 60) % 24;
  const days = Math.floor(seconds / 60 / 60 / 24);

  let ageString;
  if (days > 0) {
    ageString = `${days}d`;
  } else if (hours > 0) {
    ageString = `${hours}h`;
  } else if (minutes > 0) {
    ageString = `${minutes}m`;
  } else {
    ageString = `${seconds}s`;
  }

  return ageString;
}
