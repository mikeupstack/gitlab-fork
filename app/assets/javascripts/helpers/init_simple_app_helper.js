import Vue from 'vue';

/**
 * Initializes a component as a simple vue app, passing the necessary props. If the element
 * has a data attribute named `data-view-model`, the content of that attributed will be
 * converted from json and passed on to the component as a prop. The root component is then
 * responsible for setting up it's children, injections, and other desired features.
 *
 * @param {string} selector css selector for where to build
 * @param {Vue.component} component The Vue compoment to be built as the root of the app
 *
 * @example
 * ```html
 * <div id='#mount-here' data-view-model="{'some': 'object'}" />
 * ```
 *
 * ```javascript
 * initSimpleApp('#mount-here', MyApp)
 * ```
 *
 * This will mount MyApp as root on '#mount-here'. It will receive {'some': 'object'} as it's
 * view model prop.
 */
export const initSimpleApp = (selector, component) => {
  const element = document.querySelector(selector);

  if (!element) {
    return null;
  }

  const props = element.dataset.viewModel ? JSON.parse(element.dataset.viewModel) : {};

  return new Vue({
    el: element,
    render(h) {
      return h(component, { props });
    },
  });
};
