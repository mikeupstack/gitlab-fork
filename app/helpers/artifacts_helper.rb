# frozen_string_literal: true

module ArtifactsHelper
  def artifacts_app_data(project)
    {
      project_path: project.full_path,
      project_id: project.id,
      can_destroy_artifacts: can?(current_user, :destroy_artifacts, project).to_s,
      artifacts_management_feedback_image_path: image_path('illustrations/chat-bubble-sm.svg')
    }
  end
end
