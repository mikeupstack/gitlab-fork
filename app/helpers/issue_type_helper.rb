# frozen_string_literal: true

module IssueTypeHelper
  def issue_type_for(issue)
    return if issue.blank?

    if Feature.enabled?(:issue_type_uses_work_item_types_table)
      issue.work_item_type.base_type
    else
      issue.issue_type
    end
  end
end
