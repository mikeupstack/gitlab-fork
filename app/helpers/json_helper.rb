# frozen_string_literal: true

module JsonHelper
  # These two JSON helpers are short-form wrappers for the Gitlab::Json
  # class, which should be used in place of .to_json calls or calls to
  # the JSON class.
  def json_generate(...)
    Gitlab::Json.generate(...)
  end

  def json_parse(...)
    Gitlab::Json.parse(...)
  end
end
