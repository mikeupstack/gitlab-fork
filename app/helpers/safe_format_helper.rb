# frozen_string_literal: true

module SafeFormatHelper
  # Returns a HTML-safe string where +format+ and +args+ are escaped via
  # `html_escape` if they are not marked as HTML-safe.
  #
  # Argument +format+ must not be marked as HTML-safe via `.html_safe`.
  #
  # Example:
  #   safe_format('Some %{open}bold%{close} text.', open: '<strong>'.html_safe, close: '</strong>'.html_safe)
  #   # => 'Some <strong>bold</strong>'
  #   safe_format('See %{user_input}', user_input: '<b>bold</b>')
  #   # => 'See &lt;b&gt;bold&lt;/b&gt;
  #
  def safe_format(format, **args)
    raise ArgumentError, 'Argument `format` must not be marked as html_safe!' if format.html_safe?

    format(
      html_escape(format),
      args.transform_values { |value| html_escape(value) }
    ).html_safe
  end
end
