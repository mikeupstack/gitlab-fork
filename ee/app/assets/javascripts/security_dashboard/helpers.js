import isPlainObject from 'lodash/isPlainObject';
import { REPORT_TYPES_ALL } from 'ee/security_dashboard/store/constants';
import { convertObjectPropsToSnakeCase } from '~/lib/utils/common_utils';
import { __ } from '~/locale';

/**
 * Provided a security reports summary from the GraphQL API, this returns an array of arrays
 * representing a properly formatted report ready to be displayed in the UI. Each sub-array consists
 * of the user-friend report's name, and the summary's payload. Note that summary entries are
 * considered empty and are filtered out of the return if the payload is `null` or don't include
 * a vulnerabilitiesCount property. Report types whose name can't be matched to a user-friendly
 * name are filtered out as well.
 *
 * Take the following summary for example:
 * {
 *   containerScanning: { vulnerabilitiesCount: 123 },
 *   invalidReportType: { vulnerabilitiesCount: 123 },
 *   dast: null,
 * }
 *
 * The formatted summary would look like this:
 * [
 *   ['containerScanning', { vulnerabilitiesCount: 123 }]
 * ]
 *
 * Note that `invalidReportType` was filtered out as it can't be matched with a user-friendly name,
 * and the DAST report was omitted because it's empty (`null`).
 *
 * @param {Object} rawSummary
 * @returns {Array}
 */
export const getFormattedSummary = (rawSummary = {}) => {
  if (!isPlainObject(rawSummary)) {
    return [];
  }
  // Convert keys to snake case so they can be matched against REPORT_TYPES keys for translation
  const snakeCasedSummary = convertObjectPropsToSnakeCase(rawSummary);
  // Convert object to an array of entries to make it easier to loop through
  const summaryEntries = Object.entries(snakeCasedSummary);
  // Filter out empty entries as we don't want to display those in the summary
  const withoutEmptyEntries = summaryEntries.filter(
    ([, scanSummary]) => scanSummary?.vulnerabilitiesCount !== undefined,
  );
  // Replace keys with translations found in REPORT_TYPES if available
  const formattedEntries = withoutEmptyEntries.map(([scanType, scanSummary]) => {
    const name = REPORT_TYPES_ALL[scanType];
    return name ? [name, scanSummary] : null;
  });
  // Filter out keys that could not be matched with any translation and are thus considered invalid
  return formattedEntries.filter((entry) => entry !== null);
};

export const PROJECT_LOADING_ERROR_MESSAGE = __('An error occurred while retrieving projects.');

export default () => ({});
