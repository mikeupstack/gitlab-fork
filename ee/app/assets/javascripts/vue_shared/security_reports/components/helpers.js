import { EMPTY_BODY_MESSAGE } from './constants';

/**
 * A helper function which validates the passed
 * in body string.
 *
 * It returns an empty string if the body has explicitly
 * been passed in as an empty string, a fallback
 * message if the body is null / undefined, else
 * it will return the original body string.
 *
 * @param {String} body the body message
 *
 * @return {String} the validated body message
 */
export const bodyWithFallBack = (body) => (body === '' ? '' : body || EMPTY_BODY_MESSAGE);

/**
 * Returns a string representation of the HTTP request
 *
 * @param {Object} httpData
 * @param {String} httpData.body
 * @param {String} httpData.method
 * @param {String} httpData.url
 * @param {{name: String, value: String}[]} httpData.headers
 *
 * @returns {String}
 */
export const getHttpString = (httpData) => {
  if (!httpData) {
    return '';
  }

  const { body, headers, method, url, statusCode, reasonPhrase } = httpData;
  const methodOrStatusCode = method || statusCode;
  const urlOrReasonPhrase = url || reasonPhrase;
  const headerString = headers.map(({ name, value }) => `${name}: ${value}`).join('\n');

  return `${methodOrStatusCode} ${urlOrReasonPhrase}\n${headerString}\n\n${bodyWithFallBack(body)}`;
};

export const getCreatedIssueForVulnerability = (vulnerability) =>
  vulnerability.issue_links?.find((link) => link.link_type === 'created');

export const getDismissalTransitionForVulnerability = (vulnerability) => {
  const latestTransition = vulnerability.state_transitions?.at(-1);
  return latestTransition?.to_state.toLowerCase() === 'dismissed' ? latestTransition : null;
};
