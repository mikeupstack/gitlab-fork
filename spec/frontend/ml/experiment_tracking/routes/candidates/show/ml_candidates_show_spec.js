import { shallowMountExtended } from 'helpers/vue_test_utils_helper';
import MlCandidatesShow from '~/ml/experiment_tracking/routes/candidates/show';
import DeleteButton from '~/ml/experiment_tracking/components/delete_button.vue';
import IncubationAlert from '~/vue_shared/components/incubation/incubation_alert.vue';

describe('MlCandidatesShow', () => {
  let wrapper;

  const createWrapper = () => {
    const candidate = {
      params: [
        { name: 'Algorithm', value: 'Decision Tree' },
        { name: 'MaxDepth', value: '3' },
      ],
      metrics: [
        { name: 'AUC', value: '.55' },
        { name: 'Accuracy', value: '.99' },
      ],
      metadata: [
        { name: 'FileName', value: 'test.py' },
        { name: 'ExecutionTime', value: '.0856' },
      ],
      info: {
        iid: 'candidate_iid',
        path_to_artifact: 'path_to_artifact',
        experiment_name: 'The Experiment',
        experiment_path: 'path/to/experiment',
        status: 'SUCCESS',
        path: 'path_to_candidate',
      },
    };

    wrapper = shallowMountExtended(MlCandidatesShow, { propsData: { candidate } });
  };

  beforeEach(createWrapper);

  const findAlert = () => wrapper.findComponent(IncubationAlert);
  const findDeleteButton = () => wrapper.findComponent(DeleteButton);

  it('shows incubation warning', () => {
    expect(findAlert().exists()).toBe(true);
  });

  it('shows delete button', () => {
    expect(findDeleteButton().exists()).toBe(true);
  });

  it('passes the delete path to delete button', () => {
    expect(findDeleteButton().props('deletePath')).toBe('path_to_candidate');
  });

  it('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
