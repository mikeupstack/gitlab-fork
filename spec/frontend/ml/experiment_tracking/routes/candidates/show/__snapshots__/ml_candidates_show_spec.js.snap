// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MlCandidatesShow renders correctly 1`] = `
<div>
  <incubation-alert-stub
    featurename="Machine learning experiment tracking"
    linktofeedbackissue="https://gitlab.com/gitlab-org/gitlab/-/issues/381660"
  />
   
  <div
    class="detail-page-header gl-flex-wrap"
  >
    <div
      class="detail-page-header-body"
    >
      <h1
        class="page-title gl-font-size-h-display flex-fill"
      >
        
        Model candidate details
      
      </h1>
       
      <delete-button-stub
        actionprimarytext="Delete candidate"
        deleteconfirmationtext="Deleting this candidate will delete the associated parameters, metrics, and metadata."
        deletepath="path_to_candidate"
        modaltitle="Delete candidate?"
      />
    </div>
  </div>
   
  <table
    class="candidate-details gl-w-full"
  >
    <tbody>
      <tr
        class="divider"
      />
       
      <tr>
        <td
          class="gl-text-secondary gl-font-weight-bold"
        >
          Info
        </td>
         
        <td
          class="gl-font-weight-bold"
        >
          ID
        </td>
         
        <td>
          candidate_iid
        </td>
      </tr>
       
      <tr>
        <td />
         
        <td
          class="gl-font-weight-bold"
        >
          Status
        </td>
         
        <td>
          SUCCESS
        </td>
      </tr>
       
      <tr>
        <td />
         
        <td
          class="gl-font-weight-bold"
        >
          Experiment
        </td>
         
        <td>
          <gl-link-stub>
            The Experiment
          </gl-link-stub>
        </td>
      </tr>
       
      <tr>
        <td />
         
        <td
          class="gl-font-weight-bold"
        >
          Artifacts
        </td>
         
        <td>
          <gl-link-stub
            href="path_to_artifact"
          >
            Artifacts
          </gl-link-stub>
        </td>
      </tr>
       
      <tr
        class="divider"
      />
       
      <tr>
        <td
          class="gl-text-secondary gl-font-weight-bold"
        >
          
            Parameters
          
        </td>
         
        <td
          class="gl-font-weight-bold"
        >
          Algorithm
        </td>
         
        <td>
          Decision Tree
        </td>
      </tr>
      <tr>
        <td />
         
        <td
          class="gl-font-weight-bold"
        >
          MaxDepth
        </td>
         
        <td>
          3
        </td>
      </tr>
      <tr
        class="divider"
      />
       
      <tr>
        <td
          class="gl-text-secondary gl-font-weight-bold"
        >
          
            Metrics
          
        </td>
         
        <td
          class="gl-font-weight-bold"
        >
          AUC
        </td>
         
        <td>
          .55
        </td>
      </tr>
      <tr>
        <td />
         
        <td
          class="gl-font-weight-bold"
        >
          Accuracy
        </td>
         
        <td>
          .99
        </td>
      </tr>
      <tr
        class="divider"
      />
       
      <tr>
        <td
          class="gl-text-secondary gl-font-weight-bold"
        >
          
            Metadata
          
        </td>
         
        <td
          class="gl-font-weight-bold"
        >
          FileName
        </td>
         
        <td>
          test.py
        </td>
      </tr>
      <tr>
        <td />
         
        <td
          class="gl-font-weight-bold"
        >
          ExecutionTime
        </td>
         
        <td>
          .0856
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
